proposals/2025-12-02_controller_shared_data.md

```diff
@@ -0,0 +1,234 @@
+  * [Ability for controller to share data at
+    runtime](https://github.com/mixxxdj/mixxx/pull/12199)
+
+> TL;DR: Allow controllers to set and get data objects of arbitrary type for
+> sharing between the controller code and the engine. Think: ControlObjects of
+> arbitrary type that controllers can declare.
+
+## Why
+
+There are multiple scenarios where controllers need to share and access data
+outside the container of the controller mapping file. This includes situations
+where some controllers expose more than one USB device that need to communicate
+with each other, or when a DJ connects multiple instances of the same hardware
+to Mixxx.
+
+### Pitfalls of the current solution
+
+ControlObjects are the normal way we share data, but:
+
+* Controllers can't declare control objects.
+* Putting controller-specific data inside Mixxx itself would be bad and lead to
+  bloat.
+* Controllers need to share more types of data than just double values.
+
+## Goals
+
+Goals and use cases for the solution as proposed in [How](#how):
+
+* Namespacing: Allow different controllers to declare same-named data objects
+  without risk of collisions
+* Support controllers with screens that require communication from HID to
+  separate Bulk USB devices (Traktor S4 MK3).
+* Build a data model foundation for users with multiple instances of the same
+  controller (CDJ-2000).
+* Ensure that the API could support the features we may wish to add in the
+  future, such as global namespace, without breaking controller mappings that
+  use the API defined here.
+
+### Audience
+
+Users of modern controllers or multiple controllers will appreciate this work.
+Specifically, this work is required to fully support the Traktor S4 MK3, which
+has separate devices for the controller and the two screens (two total USB
+devices).
+
+## Non-Goals
+
+* We do not intend to fully support the multi-device scenario yet, that needs
+  further design to associate specific devices with specific controller
+  configurations.
+* We do not intend to immediately support "global" objects accessible across
+  namespaces, like "universal shift".
+
+## "Universal Shift"
+
+"Universal Shift" refers to the idea that a shift button pressed on one
+controller can be detected by any and all other controllers. This may be a
+useful use-case but creates a lot of difficulties, so for now Universal Shift is
+out of scope for this first implementation.
+
+## How
+
+### Data Object
+
+We will create a central object inside Mixxx that contains a triple-keyed map:
+
+* Namespace (string)
+  * Entity (string)
+    * Key (string)
+      * Value ("SafeData")
+
+#### Namespace
+
+`Namespace` is a string that is unique to each controller **mapping
+definition**. All connected controllers of the same model will share the same
+namespace. e.g. Two CDJ-2000's will both have a namespace like `CDJ_2000`. No
+two hardware mappings will have the same namespace, and we can enforce that with
+a precommit check.
```

Comment:
The namespace is already part of some mapping XML files now, like:
` <controller id="Traktor" namespace="S4MK3">`